### ISSUE TITLE
Difference between using a model through SentenceTransformer or with transformer directly

### ISSUE THREAD

**Issue by mapa17** [UKPLab/sentence-transformers#2261]
Hi, I am new to SentenceTransformer and would like to understand how to use it for sentence embeddings of longer texts.

I therefore tried to disentangle tokenization from model the embedding calculation and found a strong discrepancy between the case when I use SentenceTransformer to calculate embeddings, or when using the transformer library directly (which gives me access to the tokenizer).

In the code below, i calculate the sentence embeddings using "all-MiniLM-L6-v2", once with SentenceTransformer() and once with transformer.AutoModel().

I normalize the embeddings in both cases, but get very different embedding vectors, evaluating their similarity using cosine. In addition, the embedding vectors of the same model for different sentences make sense int the SentenceTransformer case, but not in the AutoModel case.

Any idea whats I am doing wrong in the AutoModel case?

```python
import os
from langchain.document_loaders import ConfluenceLoader
from langchain.embeddings import HuggingFaceEmbeddings, SentenceTransformerEmbeddings
from sentence_transformers import SentenceTransformer, util

sentences = ["Hello this is a test", "yet another short sentence", "Some longer sentence with no relevant content" + "ABC"*10]
sentences.append(sentences[0]+"!")
sentences.append(sentences[0]+"!!")
for i, s in enumerate(sentences):
    print(f"{i}: {s}")

embedding_model = SentenceTransformer('all-MiniLM-L6-v2', device='cpu')
st_embeddings = embedding_model.encode(sentences, normalize_embeddings=True, convert_to_numpy=True)
embs = st_embeddings
for i in range(len(sentences)):
    print(f"SentenceTransformer Cosine similarity: 0->{i} {util.cos_sim(embs[0], embs[i])}")

#Load AutoModel from huggingface model repository
from transformers import AutoTokenizer, AutoModel
import torch

tokenizer = AutoTokenizer.from_pretrained("sentence-transformers/all-MiniLM-L6-v2")
model = AutoModel.from_pretrained("sentence-transformers/all-MiniLM-L6-v2")

#Tokenize sentences
encoded_input = tokenizer(sentences, padding=True, truncation=True, max_length=510, return_tensors='pt')

#Compute token embeddings
with torch.no_grad():
    model_output = model(**encoded_input)
    # Normalize each vector independently to euclidian norm (p=2)
    # Note: not sure if i use the model output correctly
    embeddings = torch.nn.functional.normalize(model_output[1], p=2, dim=1)
    t_embeddings = embeddings.cpu().numpy()

embs = t_embeddings
for i in range(len(sentences)):
    print(f"Transformer Cosine similarity: 0->{i} {util.cos_sim(embs[0], embs[i])}")

for i, s in enumerate(sentences):
    print(f"Cosine similarity between st and t embeddings: {i}-{i} {util.cos_sim(st_embeddings[i], t_embeddings[i])}")
```

Gives the result

```bash
0: Hello this is a test
1: yet another short sentence
2: Some longer sentence with no relevant contentABCABCABCABCABCABCABCABCABCABC
3: Hello this is a test!
4: Hello this is a test!!
SentenceTransformer Cosine similarity: 0->0 tensor([[1.0000]])
SentenceTransformer Cosine similarity: 0->1 tensor([[0.3102]])
SentenceTransformer Cosine similarity: 0->2 tensor([[0.1638]])
SentenceTransformer Cosine similarity: 0->3 tensor([[0.9179]])
SentenceTransformer Cosine similarity: 0->4 tensor([[0.8582]])
Transformer Cosine similarity: 0->0 tensor([[1.]])
Transformer Cosine similarity: 0->1 tensor([[0.7652]])
Transformer Cosine similarity: 0->2 tensor([[0.7532]])
Transformer Cosine similarity: 0->3 tensor([[0.9940]])
Transformer Cosine similarity: 0->4 tensor([[0.9897]])
Cosine similarity between st and t embeddings: 0-0 tensor([[0.0021]])
Cosine similarity between st and t embeddings: 1-1 tensor([[-0.1045]])
Cosine similarity between st and t embeddings: 2-2 tensor([[-0.0345]])
Cosine similarity between st and t embeddings: 3-3 tensor([[0.0285]])
Cosine similarity between st and t embeddings: 4-4 tensor([[0.0493]])
```

**Comment by carlesoctav** [comment]
I believe that when using the Hugging Face AutoModel interface with an all-lm model, the output is still a word embedding for each token. You need to average them, as shown in the example on the [page for the all-MiniLM-L6-v2 model](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2).


```python
def mean_pooling(model_output, attention_mask):
    token_embeddings = model_output[0] # The first element of model_output contains all token embeddings.
    input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
    return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)

```

**Comment by mapa17** [comment]
Thank you, yes I only afterwards realized that the SentenceTransformer models are always SequenceModels that include the original tokenizer plus a pooling layer.
